lib/TH/generic/THTensorRandom.c

```diff
@@ -70,6 +70,141 @@ void THTensor_(logNormal)(THTensor *self, THGenerator *_generator, double mean,
   TH_TENSOR_APPLY(real, self, *self_data = (real)THRandom_logNormal(_generator, mean, stdv););
 }
 
+
+void THTensor_(alias_multinomial_setup)(THTensor *probs, THLongTensor *J, THTensor *q)
+{
+  long inputsize = THTensor_(nElement)(probs);
+  long i = 0;
+  THLongTensor *smaller = THLongTensor_newWithSize1d(inputsize);
+  THLongTensor *larger = THLongTensor_newWithSize1d(inputsize);
+  long small_c = 0;
+  long large_c = 0;
+  THLongTensor_resize1d(J, inputsize);
+  THTensor_(resize1d)(q, inputsize);
+  
+  for(i = 0; i < inputsize; i++)
+    {
+      THLongStorage_set(J->storage,
+                        J->storageOffset+i*J->stride[0],
+                        0L);
+      
+      real val = THStorage_(get)(probs->storage,
+                                 probs->storageOffset+i*probs->stride[0]);
+
+      THStorage_(set)(q->storage,
+                      q->storageOffset+i*q->stride[0],
+                      inputsize*val);
+
+      
+      if (inputsize * val < 1.0)
+        {
+          THLongStorage_set(smaller->storage,
+                            smaller->storageOffset+small_c*smaller->stride[0],
+                            i);
+          small_c += 1;
+        }
+      else
+        {
+          THLongStorage_set(larger->storage,
+                            larger->storageOffset+large_c*larger->stride[0],
+                            i);
+          large_c += 1;
+        }
+    }
+
+  // Loop through and create little binary mixtures that
+  // appropriately allocate the larger outcomes over the
+  // overall uniform mixture.
+  long large, small;
+  while(small_c > 0 && large_c > 0)
+    {
+      large = THLongStorage_get(larger->storage,
+                                larger->storageOffset+(large_c-1)*larger->stride[0]);
+      small = THLongStorage_get(smaller->storage,
+                                smaller->storageOffset+(small_c-1)*smaller->stride[0]);
+      
+      THLongStorage_set(J->storage,
+                        J->storageOffset+small*J->stride[0],
+                        large+1L);
```

Comment:
This can be replaced with `THLongTensor_fastSet(J, small, large+1L)`. Same applies to most other places. There's no need to unpack the storage from the tensor